auto_submit/lib/requests/check_pull_request.dart

```diff
@@ -27,23 +27,52 @@ class CheckPullRequest extends RequestHandler {
     final String rawBody = await request.readAsString();
     final body = json.decode(rawBody) as Map<String, dynamic>;
     final PullRequest pullRequest = PullRequest.fromJson(body['pull_request']);
+    final RepositorySlug slug = pullRequest.base!.repo!.slug();
 
     final GithubService gitHub = await config.createGithubService();
     final _AutoMergeQueryResult queryResult = await _parseQueryData(pullRequest, gitHub);
-    if (await shouldMergePullRequest(queryResult)) {
+    if (await shouldMergePullRequest(queryResult, slug, gitHub)) {
       // TODO(Kristin): Keep pulling pubsub queue, https://github.com/flutter/flutter/issues/98704
 
-    } else {
+    } else if (queryResult.shouldRemoveLabel) {
+      log.info('Removing label for commit: ${queryResult.sha}');
+      await _removeLabel();
       return Response.ok(jsonEncode(<String, String>{}));
     }
 
     return Response.ok(rawBody);
   }
 
   /// Check if the pull request should be merged.
-  Future<bool> shouldMergePullRequest(_AutoMergeQueryResult queryResult) async {
-    // TODO(Kristin): Implement shouldMergePullRequest, https://github.com/flutter/flutter/issues/98707
+  Future<bool> shouldMergePullRequest(
+      _AutoMergeQueryResult queryResult, RepositorySlug slug, GithubService github) async {
+    if (queryResult.shouldMerge) {
+      return true;
+    }
+    // If the PR is a revert of the tot commit, merge without waiting for checks passing.
+    return await isTOTRevert(queryResult.sha, slug, github);
+  }
+
+  /// Check if the `commitSha` is a clean revert of TOT commit.
+  ///
+  /// By comparing the current commit with second TOT commit, an empty `files` in
+  /// `GitHubComparison` validates a clean revert of TOT commit.
+  ///
+  /// Note: [compareCommits] expects base commit first, and then head commit.
+  Future<bool> isTOTRevert(String headSha, RepositorySlug slug, GithubService github) async {
+    final RepositoryCommit secondTotCommit = await github.getRepoCommit(slug, 'HEAD~');
```

Comment:
Why are we using `HEAD~` instead of `HEAD`?

We want to compare the pullRequest.base.sha to the repositories HEAD

Reply:
Hi Casey, I have the same doubts before, i also thought here should be HEAD. However, I noticed that the previous logic https://github.com/flutter/cocoon/blob/main/app_dart/lib/src/request_handlers/check_for_waiting_pull_requests.dart#L117 here used the HEAD~, I thought this was a special case so I just retained what he used.

Reply:
Can we add clear documentation? I remember this being confusing last time I reviewed it, and I'm not sure I still understand. It's an issue with the compare files, right?

Here's an example I did against flutter/flutter that shows why I would expect HEAD:

```sh
$ git log -n 2
* 57aa49028d (HEAD -> master, upstream/master) Migrate NavigationRail to Material 3. (#99171)  Darren Austin, 79 minutes ago
* 9291c8ba31 Remove expired ThemeData deprecations (#98578)  Kate Lovett, 80 minutes ago
$ git log HEAD -n 1
* 57aa49028d (HEAD -> master, upstream/master) Migrate NavigationRail to Material 3. (#99171)  Darren Austin, 79 minutes ago
$ git log HEAD~ -n 1
* 9291c8ba31 Remove expired ThemeData deprecations (#98578)  Kate Lovett, 80 minutes ago
```

Reply:
To clarify, there is no issue with the compareCommits API. Let's take the provided 2 commits for example:
If we have a clean revert of `57aa49028d`, and suppose the revert has commit sha `abcdefg`:
1) `files` of `github.repositories.compareCommits(slug, '57aa49028d', 'abcdefg')` contains all the changed files from `57aa49028d`, but in the opposite way (`+` => `-`, and `-` => `+)
2) files` of `github.repositories.compareCommits(slug, '9291c8ba31', 'abcdefg')` will be empty, as revert commit `abcdefg` reverts all changes from the TOT `57aa49028d`. And this is what we want to validate.